File: lib/token_lens/renderer/layout.rb
# frozen_string_literal: true

module TokenLens
  module Renderer
    class Layout
      CANVAS_WIDTH = 1200
      ROW_HEIGHT = 32
      MIN_THREAD_WIDTH = 80  # minimum pixels per root thread so slivers are readable

      def initialize(canvas_width: CANVAS_WIDTH)
        @canvas_width = canvas_width
      end

      def layout(nodes)
        max_depth = all_nodes(nodes).map { |n| n[:depth] }.max || 0
        effective_width = [@canvas_width, nodes.length * MIN_THREAD_WIDTH].max

        total = nodes.sum { |n| n[:subtree_tokens] }
        scale = (total > 0) ? effective_width.to_f / total : 1.0
        position(nodes, x: 0, scale: scale, max_depth: max_depth)

        total_cost = nodes.sum { |n| n[:subtree_cost] }
        cost_scale = (total_cost > 0) ? effective_width.to_f / total_cost : 1.0
        position_cost(nodes, x: 0, scale: cost_scale, max_depth: max_depth)

        effective_width
      end

      private

      # Bottom-up layout: roots at bottom (y = max_depth * ROW_HEIGHT),
      # deepest children at top (y = 0).
      def position(nodes, x:, scale:, max_depth:)
        cursor = x.to_f
        nodes.each do |node|
          start = cursor.round
          cursor += node[:subtree_tokens] * scale
          node[:x] = start
          node[:y] = (max_depth - node[:depth]) * ROW_HEIGHT
          node[:w] = cursor.round - start
          position(node[:children], x: node[:x], scale: scale, max_depth: max_depth)
        end
      end

      def position_cost(nodes, x:, scale:, max_depth:)
        cursor = x.to_f
        nodes.each do |node|
          start = cursor.round
          cursor += node[:subtree_cost] * scale
          node[:cost_x] = start
          node[:cost_w] = cursor.round - start
          position_cost(node[:children], x: node[:cost_x], scale: scale, max_depth: max_depth)
        end
      end

      def all_nodes(nodes)
        nodes.flat_map { |n| [n, *all_nodes(n[:children])] }
      end
    end
  end
end

File: src/renderer/layout.ts
import type { Node } from "../types";

const CANVAS_WIDTH = 1200;
const ROW_HEIGHT = 32;
const MIN_THREAD_WIDTH = 80;

export function layout(nodes: Node[]): number {
  const maxDepth = allNodes(nodes).reduce((max, n) => Math.max(max, n.depth ?? 0), 0);
  const effectiveWidth = Math.max(CANVAS_WIDTH, nodes.length * MIN_THREAD_WIDTH);

  const total = nodes.reduce((sum, n) => sum + (n.subtreeTokens ?? 0), 0);
  const scale = total > 0 ? effectiveWidth / total : 1.0;
  position(nodes, 0, scale, maxDepth);

  const totalCost = nodes.reduce((sum, n) => sum + (n.subtreeCost ?? 0), 0);
  const costScale = totalCost > 0 ? effectiveWidth / totalCost : 1.0;
  positionCost(nodes, 0, costScale, maxDepth);

  return effectiveWidth;
}

function position(nodes: Node[], x: number, scale: number, maxDepth: number): void {
  let cursor = x;
  for (const node of nodes) {
    const start = Math.round(cursor);
    cursor += (node.subtreeTokens ?? 0) * scale;
    node.x = start;
    node.y = (maxDepth - (node.depth ?? 0)) * ROW_HEIGHT;
    node.w = Math.round(cursor) - start;
    position(node.children, node.x, scale, maxDepth);
  }
}

function positionCost(nodes: Node[], x: number, scale: number, maxDepth: number): void {
  let cursor = x;
  for (const node of nodes) {
    const start = Math.round(cursor);
    cursor += (node.subtreeCost ?? 0) * scale;
    node.costX = start;
    node.costW = Math.round(cursor) - start;
    positionCost(node.children, node.costX, scale, maxDepth);
  }
}

function allNodes(nodes: Node[]): Node[] {
  return nodes.flatMap((n) => [n, ...allNodes(n.children)]);
}
